Helm lint (v3) doesn't take values files into account during schema validation

Edit: Steps to reproduce: https://github.com/helm/helm/issues/7756#issuecomment-617889123

I have a bunch of charts with `values.schema.json` files that won't pass schema validation without passing in necessary, environment-specific values. I don't want these values in the base `values.yaml` file because they are environment-specific, but should fail at install/upgrade time.

When running `helm lint ${chart} --values ./lint-values.yaml` I'm getting errors like below **even though these values are provided in `./lint-values.yaml`**:
```
==> Linting ${chart}
[ERROR] values.yaml:
- ingress.hosts: Array must have at least 1 items
...

Error: 1 chart(s) linted, 1 chart(s) failed
```

Important to note that `helm template ${chart} --values ./line-values.yaml` does not throw any schema validation errors. When I run `helm template ${chart}` I see the same errors as above:
```
Error: values don't meet the specifications of the schema(s) in the following chart(s):
agent-services:
- ingress.hosts: Array must have at least 1 items
...
```

I think this is a bug. `helm lint` should run values schema validation against the final `.Values` object like the docs say:
> Note that the schema is applied to the final `.Values` object, and not just to the `values.yaml` file. This means that the following `yaml` file is valid, given that the chart is installed with the appropriate `--set` option shown below.

Source: https://helm.sh/docs/topics/charts/#schema-files

---

Output of `helm version`:
```
version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

Output of `kubectl version`: not relevant

Cloud Provider/Platform (AKS, GKE, Minikube etc.): not relevant

Also ran into this, but I'm not sure (at least in my case) that it has to do with values files not being taken into account.  It looks like there's two passes and provided values are only being used in one of them.

I'm trying json schema validation for the first time on a new chart and I'm seeing that it's outputting errors _twice_.  I've also tossed in a "bug" (`{{ asdf }}`) in one of the templates which should output a failure.

```
$ helm lint test/
==> Linting test/
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - app.baseUri: Does not match pattern '^(https|http|)://'
- app.baseUri: Does not match format 'uri'
- oidc.authority: Does not match pattern '^(https|http|)://'
- oidc.authority: Does not match format 'uri'
- appVersion: Does not match pattern '^\d((\.\d+)?)+'
- api.baseUri: Does not match pattern '^(https|http)://.+'
- api.baseUri: Does not match format 'uri'

[ERROR] templates/: values don't meet the specifications of the schema(s) in the following chart(s):
test:
- appVersion: Does not match pattern '^\d((\.\d+)?)+'
- api.baseUri: Does not match pattern '^(https|http)://.+'
- api.baseUri: Does not match format 'uri'
- app.baseUri: Does not match pattern '^(https|http|)://'
- app.baseUri: Does not match format 'uri'
- oidc.authority: Does not match pattern '^(https|http|)://'
- oidc.authority: Does not match format 'uri'


Error: 1 chart(s) linted, 1 chart(s) failed
```
Note that the errors are spit out twice, and no template validation error.

Now if I specify a values file (or specify them all via the command line with `--set`, same result):
```
$ helm lint test/ -f .testvalues.yaml 
==> Linting test/
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - appVersion: Does not match pattern '^\d((\.\d+)?)+'
- api.baseUri: Does not match pattern '^(https|http)://.+'
- api.baseUri: Does not match format 'uri'
- app.baseUri: Does not match pattern '^(https|http|)://'
- app.baseUri: Does not match format 'uri'
- oidc.authority: Does not match pattern '^(https|http|)://'
- oidc.authority: Does not match format 'uri'

[ERROR] templates/: parse error at (test/templates/deployment.yaml:98): function "asdf" not defined

Error: 1 chart(s) linted, 1 chart(s) failed
```

Here the errors are only printed once, and it looks like in what I'm assuming is a "second validation pass", schema validation succeeded and it went on to validate the templates, showing the `function "asdf" not defined` error.  So, it seems the values file was only taken into account in the second pass, but not the first.

Ideally we would only have one pass, based on the final values object.  It would also be great if the linting didn't stop if any schema validations occur, and proceeded with validating the chart - I didn't see that behavior described in documentation, and the reason why I'm using a `.testvalues.yaml` to provide valid values so that I can see the template validation errors.

--- a/internal/test/ensure/ensure.go
+++ b/internal/test/ensure/ensure.go
@@ -19,6 +19,7 @@ package ensure
 import (
 	"io/ioutil"
 	"os"
+	"path/filepath"
 	"testing"
 
 	"helm.sh/helm/v3/pkg/helmpath/xdg"
@@ -44,4 +45,22 @@ func TempDir(t *testing.T) string {
 		t.Fatal(err)
 	}
 	return d
+}
+
+// TempFile ensures a temp file for unit testing purposes.
+//
+// It returns the path to the directory (to which you will still need to join the filename)
+//
+// You must clean up the directory that is returned.
+//
+// 	tempdir := TempFile(t, "foo", []byte("bar"))
+// 	defer os.RemoveAll(tempdir)
+// 	filename := filepath.Join(tempdir, "foo")
+func TempFile(t *testing.T, name string, data []byte) string {
+	path := TempDir(t)
+	filename := filepath.Join(path, name)
+	if err := ioutil.WriteFile(filename, data, 0755); err != nil {
+		t.Fatal(err)
+	}
+	return path
 }

--- a/pkg/lint/lint.go
+++ b/pkg/lint/lint.go
@@ -30,7 +30,7 @@ func All(basedir string, values map[string]interface{}, namespace string, strict
 
 	linter := support.Linter{ChartDir: chartDir}
 	rules.Chartfile(&linter)
-	rules.Values(&linter)
+	rules.ValuesWithOverrides(&linter, values)
 	rules.Templates(&linter, values, namespace, strict)
 	return linter
 }

--- a/pkg/lint/rules/values.go
+++ b/pkg/lint/rules/values.go
@@ -36,7 +36,7 @@ func Values(linter *support.Linter) {
 
 // ValuesWithOverrides tests the values.yaml file.
 //
-// If a schema iis present in the chart, values are tested against that. Otherwise,
+// If a schema is present in the chart, values are tested against that. Otherwise,
 // they are only tested for well-formedness.
 //
 // If additional values are supplied, they are coalesced into the values in values.yaml.

--- a/pkg/lint/rules/values.go
+++ b/pkg/lint/rules/values.go
@@ -28,7 +28,19 @@ import (
 )
 
 // Values lints a chart's values.yaml file.
+//
+// This function is deprecated and will be removed in Helm 4.
 func Values(linter *support.Linter) {
+	ValuesWithOverrides(linter, map[string]interface{}{})
+}
+
+// ValuesWithOverrides tests the values.yaml file.
+//
+// If a schema iis present in the chart, values are tested against that. Otherwise,
+// they are only tested for well-formedness.
+//
+// If additional values are supplied, they are coalesced into the values in values.yaml.
+func ValuesWithOverrides(linter *support.Linter, values map[string]interface{}) {
 	file := "values.yaml"
 	vf := filepath.Join(linter.ChartDir, file)
 	fileExists := linter.RunLinterRule(support.InfoSev, file, validateValuesFileExistence(vf))
@@ -37,7 +49,7 @@ func Values(linter *support.Linter) {
 		return
 	}
 
-	linter.RunLinterRule(support.ErrorSev, file, validateValuesFile(vf))
+	linter.RunLinterRule(support.ErrorSev, file, validateValuesFile(vf, values))
 }
 
 func validateValuesFileExistence(valuesPath string) error {
@@ -48,11 +60,18 @@ func validateValuesFileExistence(valuesPath string) error {
 	return nil
 }
 
-func validateValuesFile(valuesPath string) error {
+func validateValuesFile(valuesPath string, overrides map[string]interface{}) error {
 	values, err := chartutil.ReadValuesFile(valuesPath)
 	if err != nil {
 		return errors.Wrap(err, "unable to parse YAML")
 	}
+
+	// Helm 3.0.0 carried over the values linting from Helm 2.x, which only tests the top
+	// level values against the top-level expectations. Subchart values are not linted.
+	// We could change that. For now, though, we retain that strategy, and thus can
+	// coalesce tables (like reuse-values does) instead of doing the full chart
+	// CoalesceValues.
+	values = chartutil.CoalesceTables(values, overrides)
 
 	ext := filepath.Ext(valuesPath)
 	schemaPath := valuesPath[:len(valuesPath)-len(ext)] + ".schema.json"

--- a/pkg/lint/rules/values_test.go
+++ b/pkg/lint/rules/values_test.go
@@ -17,14 +17,39 @@ limitations under the License.
 package rules
 
 import (
+	"io/ioutil"
 	"os"
 	"path/filepath"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"helm.sh/helm/v3/internal/test/ensure"
 )
 
-var (
-	nonExistingValuesFilePath = filepath.Join("/fake/dir", "values.yaml")
-)
+var nonExistingValuesFilePath = filepath.Join("/fake/dir", "values.yaml")
+
+const testSchema = `
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "helm values test schema",
+  "type": "object",
+  "additionalProperties": false,
+  "required": [
+    "username",
+    "password"
+  ],
+  "properties": {
+    "username": {
+      "description": "Your username",
+      "type": "string"
+    },
+    "password": {
+      "description": "Your password",
+      "type": "string"
+    }
+  }
+}
+`
 
 func TestValidateValuesYamlNotDirectory(t *testing.T) {
 	_ = os.Mkdir(nonExistingValuesFilePath, os.ModePerm)
@@ -34,4 +59,69 @@ func TestValidateValuesYamlNotDirectory(t *testing.T) {
 	if err == nil {
 		t.Errorf("validateValuesFileExistence to return a linter error, got no error")
 	}
+}
+
+func TestValidateValuesFileWellFormed(t *testing.T) {
+	badYaml := `
+	not:well[]{}formed
+	`
+	tmpdir := ensure.TempFile(t, "values.yaml", []byte(badYaml))
+	defer os.RemoveAll(tmpdir)
+	valfile := filepath.Join(tmpdir, "values.yaml")
+	if err := validateValuesFile(valfile, map[string]interface{}{}); err == nil {
+		t.Fatal("expected values file to fail parsing")
+	}
+}
+
+func TestValidateValuesFileSchema(t *testing.T) {
+	yaml := "username: admin\npassword: swordfish"
+	tmpdir := ensure.TempFile(t, "values.yaml", []byte(yaml))
+	defer os.RemoveAll(tmpdir)
+	createTestingSchema(t, tmpdir)
+
+	valfile := filepath.Join(tmpdir, "values.yaml")
+	if err := validateValuesFile(valfile, map[string]interface{}{}); err != nil {
+		t.Fatalf("Failed validation with %s", err)
+	}
+}
+
+func TestValidateValuesFileSchemaFailure(t *testing.T) {
+	// 1234 is an int, not a string. This should fail.
+	yaml := "username: 1234\npassword: swordfish"
+	tmpdir := ensure.TempFile(t, "values.yaml", []byte(yaml))
+	defer os.RemoveAll(tmpdir)
+	createTestingSchema(t, tmpdir)
+
+	valfile := filepath.Join(tmpdir, "values.yaml")
+
+	err := validateValuesFile(valfile, map[string]interface{}{})
+	if err == nil {
+		t.Fatal("expected values file to fail parsing")
+	}
+
+	assert.Contains(t, err.Error(), "Expected: string, given: integer", "integer should be caught by schema")
+}
+
+func TestValidateValuesFileSchemaOverrides(t *testing.T) {
+	yaml := "username: admin"
+	overrides := map[string]interface{}{
+		"password": "swordfish",
+	}
+	tmpdir := ensure.TempFile(t, "values.yaml", []byte(yaml))
+	defer os.RemoveAll(tmpdir)
+	createTestingSchema(t, tmpdir)
+
+	valfile := filepath.Join(tmpdir, "values.yaml")
+	if err := validateValuesFile(valfile, overrides); err != nil {
+		t.Fatalf("Failed validation with %s", err)
+	}
+}
+
+func createTestingSchema(t *testing.T, dir string) string {
+	t.Helper()
+	schemafile := filepath.Join(dir, "values.schema.json")
+	if err := ioutil.WriteFile(schemafile, []byte(testSchema), 0700); err != nil {
+		t.Fatalf("Failed to write schema to tmpdir: %s", err)
+	}
+	return schemafile
 }

--- a/pkg/lint/rules/values_test.go
+++ b/pkg/lint/rules/values_test.go
@@ -23,6 +23,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+
 	"helm.sh/helm/v3/internal/test/ensure"
 )
 

